drivers/usb/usbstor_new/fdo.c

```diff
@@ -459,3 +486,149 @@ USBSTOR_FdoHandlePnp(
     //
     return Status;
 }
+
+VOID
+NTAPI
+USBSTOR_FdoSetPowerCompletion(
+    IN PDEVICE_OBJECT DeviceObject,
+    IN UCHAR MinorFunction,
+    IN POWER_STATE PowerState,
+    IN PVOID Context,
+    IN PIO_STATUS_BLOCK IoStatus)
+{
+    PDEVICE_OBJECT FdoDevice;
+    PFDO_DEVICE_EXTENSION FdoExtension;
+    PIRP CurrentPowerIrp;
+
+    FdoDevice = Context;
+    FdoExtension = FdoDevice->DeviceExtension;
+
+    CurrentPowerIrp = FdoExtension->CurrentPowerIrp;
+    FdoExtension->CurrentPowerIrp = NULL;
+
+    PoStartNextPowerIrp(CurrentPowerIrp);
+
+    IoCopyCurrentIrpStackLocationToNext(CurrentPowerIrp);
+    IoMarkIrpPending(CurrentPowerIrp);
+
+    PoCallDriver(FdoExtension->LowerDeviceObject, CurrentPowerIrp);
+}
+
+NTSTATUS
+NTAPI
+USBSTOR_FdoSetD0Completion(
+    IN PDEVICE_OBJECT DeviceObject,
+    IN PIRP Irp,
+    IN PVOID Context)
+{
+    NTSTATUS Status;
+    KIRQL OldIrql;
+
+    Status = Irp->IoStatus.Status;
+
+    KeRaiseIrql(DISPATCH_LEVEL, &OldIrql);
+    IoStartNextPacket(DeviceObject, TRUE);
+    KeLowerIrql(OldIrql);
+
+    PoStartNextPowerIrp(Irp);
+    return Status;
+}
+
+NTSTATUS
+NTAPI
+USBSTOR_FdoSetPower(
+    IN PDEVICE_OBJECT DeviceObject,
+    IN PIRP Irp)
+{
+    PFDO_DEVICE_EXTENSION FdoExtension;
+    PIO_STACK_LOCATION IoStack;
+    POWER_STATE_TYPE Type;
+    POWER_STATE State;
+    POWER_STATE NewState;
+    POWER_STATE OldState;
+
+    DPRINT("USBSTOR_FdoSetPower: DeviceObject %p, Irp %p\n", DeviceObject, Irp);
+
+    FdoExtension = DeviceObject->DeviceExtension;
+
+    IoStack = IoGetCurrentIrpStackLocation(Irp);
+    Type = IoStack->Parameters.Power.Type;
+    State = IoStack->Parameters.Power.State;
+
+    switch (Type)
+    {
+        case SystemPowerState:
```

Comment:
As these 'case's are huge, maybe it would be nice to put their code inside {} blocks.

Reply:
Added